File: packages/nerva-js/src/router/embedding.ts
/**
 * Embedding-based router — cosine similarity against handler descriptions.
 *
 * Routes messages by embedding the input text and comparing it against
 * pre-computed embeddings of handler descriptions. Top-k handlers above
 * the confidence threshold are returned. No LLM call needed.
 *
 * @module router/embedding
 */

import type { ExecContext } from "../context.js";
import type { HandlerCandidate, IntentResult, IntentRouter } from "./index.js";
import { createHandlerCandidate, createIntentResult } from "./index.js";

// ---------------------------------------------------------------------------
// Native module auto-detection
// ---------------------------------------------------------------------------

/** Shape of the @nerva/core native addon. */
interface NativeModule {
  cosine_similarity(a: number[], b: number[]): number;
}

/**
 * Attempt to load the native Rust module at startup.
 * Falls back to null when the addon is not installed or not compiled
 * for the current platform — pure JS takes over transparently.
 */
let nativeModule: NativeModule | null = null;
try {
  // eslint-disable-next-line @typescript-eslint/no-require-imports
  nativeModule = require("@nerva/core") as NativeModule;
} catch {
  // Native module not available — pure JS fallback is used.
}

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

/** Dense vector representation of a text string. */
export type Embedding = readonly number[];

/**
 * Async function that converts text into an embedding vector.
 *
 * Implementations may call an external API (OpenAI, Cohere) or run
 * a local model. The returned vector length must be consistent
 * across calls.
 */
export type EmbedFn = (text: string) => Promise<number[]>;

// ---------------------------------------------------------------------------
// Constants
// ---------------------------------------------------------------------------

/** Minimum valid threshold value. */
const MINIMUM_THRESHOLD = 0.0;

/** Maximum valid threshold value. */
const MAXIMUM_THRESHOLD = 1.0;

/** Default cosine similarity threshold to consider a match. */
const DEFAULT_THRESHOLD = 0.3;

/** Default number of top candidates to return. */
const DEFAULT_TOP_K = 5;

/** Confidence for empty results. */
const NO_MATCH_CONFIDENCE = 0.0;

/** Intent label when nothing matched. */
const NO_MATCH_INTENT = "unknown";

/** Intent label for semantic matches. */
const MATCH_INTENT = "semantic";

// ---------------------------------------------------------------------------
// Internal value object
// ---------------------------------------------------------------------------

/** A handler with its pre-computed description embedding. */
interface HandlerDescriptor {
  readonly name: string;
  readonly description: string;
  readonly embedding: number[];
}

// ---------------------------------------------------------------------------
// EmbeddingRouter
// ---------------------------------------------------------------------------

/** Configuration options for {@link EmbeddingRouter}. */
export interface EmbeddingRouterOptions {
  /** Minimum cosine similarity to consider a match (0.0-1.0). */
  readonly threshold?: number;
  /** Maximum number of candidates to return. */
  readonly topK?: number;
}

/**
 * Router using cosine similarity between message and handler descriptions.
 *
 * Handlers are registered with descriptions. At classify time the message
 * is embedded and compared against all handler embeddings. Top-k handlers
 * above the confidence threshold are returned, best first.
 */
export class EmbeddingRouter implements IntentRouter {
  private readonly _embed: EmbedFn;
  private readonly _threshold: number;
  private readonly _topK: number;
  private readonly _handlers: HandlerDescriptor[] = [];

  /**
   * @param embed - Async function that converts text to an embedding vector.
   * @param options - Optional threshold and topK overrides.
   * @throws {RangeError} If threshold is outside [0.0, 1.0].
   * @throws {RangeError} If topK is less than 1.
   */
  constructor(embed: EmbedFn, options?: EmbeddingRouterOptions) {
    const threshold = options?.threshold ?? DEFAULT_THRESHOLD;
    const topK = options?.topK ?? DEFAULT_TOP_K;

    if (threshold < MINIMUM_THRESHOLD || threshold > MAXIMUM_THRESHOLD) {
      throw new RangeError(
        `threshold must be between ${MINIMUM_THRESHOLD} and ${MAXIMUM_THRESHOLD}, got ${threshold}`,
      );
    }
    if (topK < 1) {
      throw new RangeError(`topK must be >= 1, got ${topK}`);
    }

    this._embed = embed;
    this._threshold = threshold;
    this._topK = topK;
  }

  /**
   * Register a handler by embedding its description.
   *
   * The description is embedded immediately via the injected embed
   * function and stored for later comparison.
   *
   * @param name - Handler name (must match a registry entry).
   * @param description - Human-readable description of what the handler does.
   * @throws {Error} If name is empty or description is blank.
   */
  async register(name: string, description: string): Promise<void> {
    if (!name) {
      throw new Error("Handler name must not be empty");
    }
    if (!description.trim()) {
      throw new Error("Handler description must not be blank");
    }

    const embedding = await this._embed(description);
    this._handlers.push({ name, description, embedding });
  }

  /**
   * Classify a message by cosine similarity against handler descriptions.
   *
   * @param message - Raw user message text.
   * @param _ctx - Execution context (carried for protocol conformance).
   * @returns IntentResult with ranked candidates.
   */
  async classify(message: string, _ctx: ExecContext): Promise<IntentResult> {
    if (!message.trim() || this._handlers.length === 0) {
      return emptyResult();
    }

    const queryEmbedding = await this._embed(message);
    const candidates = rankHandlers(
      queryEmbedding,
      this._handlers,
      this._threshold,
      this._topK,
    );

    if (candidates.length === 0) {
      return emptyResult();
    }

    return buildResult(candidates);
  }
}

// ---------------------------------------------------------------------------
// Pure helpers
// ---------------------------------------------------------------------------

/**
 * Compute cosine similarity between two vectors.
 *
 * Returns 0.0 for zero-length vectors or mismatched dimensions
 * rather than raising, so callers never hit division-by-zero.
 *
 * @param a - First embedding vector.
 * @param b - Second embedding vector.
 * @returns Cosine similarity in [-1.0, 1.0], or 0.0 for degenerate inputs.
 */
export function cosineSimilarity(a: Embedding, b: Embedding): number {
  // Delegate to native Rust SIMD implementation when available.
  if (nativeModule) {
    return nativeModule.cosine_similarity([...a], [...b]);
  }

  return cosineSimilarityJS(a, b);
}

/**
 * Pure JavaScript cosine similarity implementation.
 *
 * Used as fallback when the native @nerva/core addon is not available.
 *
 * @param a - First embedding vector.
 * @param b - Second embedding vector.
 * @returns Cosine similarity in [-1.0, 1.0], or 0.0 for degenerate inputs.
 */
function cosineSimilarityJS(a: Embedding, b: Embedding): number {
  if (a.length !== b.length || a.length === 0) {
    return 0.0;
  }

  let dot = 0;
  let normA = 0;
  let normB = 0;

  for (let i = 0; i < a.length; i++) {
    const ai = a[i]!;
    const bi = b[i]!;
    dot += ai * bi;
    normA += ai * ai;
    normB += bi * bi;
  }

  const magA = Math.sqrt(normA);
  const magB = Math.sqrt(normB);

  if (magA === 0 || magB === 0) {
    return 0.0;
  }

  return dot / (magA * magB);
}

/**
 * Score all handlers against the query embedding and return top-k above threshold.
 *
 * @param queryEmbedding - Embedded user message.
 * @param handlers - All registered handler descriptors.
 * @param threshold - Minimum similarity to include.
 * @param topK - Maximum candidates to return.
 * @returns Sorted candidates (best first), capped at topK.
 */
function rankHandlers(
  queryEmbedding: number[],
  handlers: readonly HandlerDescriptor[],
  threshold: number,
  topK: number,
): HandlerCandidate[] {
  const scored: Array<{ score: number; descriptor: HandlerDescriptor }> = [];

  for (const handler of handlers) {
    const similarity = cosineSimilarity(queryEmbedding, handler.embedding);
    const clamped = Math.max(0.0, Math.min(1.0, similarity));
    if (clamped >= threshold) {
      scored.push({ score: clamped, descriptor: handler });
    }
  }

  scored.sort((a, b) => b.score - a.score);

  return scored.slice(0, topK).map(({ score, descriptor }) =>
    createHandlerCandidate(
      descriptor.name,
      score,
      `Cosine similarity ${score.toFixed(4)} with '${descriptor.description}'`,
    ),
  );
}

/**
 * Build an IntentResult from ranked candidates.
 *
 * @param candidates - Non-empty list of ranked handler candidates.
 * @returns IntentResult with semantic intent and top-candidate confidence.
 */
function buildResult(candidates: readonly HandlerCandidate[]): IntentResult {
  const rawScores: Record<string, number> = {};
  for (const c of candidates) {
    rawScores[c.name] = c.score;
  }
  const topCandidate = candidates[0];
  const confidence = topCandidate !== undefined ? topCandidate.score : NO_MATCH_CONFIDENCE;
  return createIntentResult(MATCH_INTENT, confidence, candidates, rawScores);
}

/**
 * Build an empty IntentResult when nothing matched.
 *
 * @returns IntentResult with zero confidence and no handlers.
 */
function emptyResult(): IntentResult {
  return createIntentResult(NO_MATCH_INTENT, NO_MATCH_CONFIDENCE, []);
}

File: packages/nerva-py/nerva/router/embedding.py
"""Embedding-based router — cosine similarity against handler descriptions.

Routes messages by embedding the input text and comparing it against
pre-computed embeddings of handler descriptions.  Top-k handlers above
the confidence threshold are returned.  No LLM call needed.
"""

from __future__ import annotations

import math
from dataclasses import dataclass
from typing import TYPE_CHECKING, Protocol

from nerva.router import HandlerCandidate, IntentResult

try:
    from nerva_core import cosine_similarity as _native_cosine
    _USE_NATIVE = True
except ImportError:
    _USE_NATIVE = False

if TYPE_CHECKING:
    from nerva.context import ExecContext

__all__ = [
    "EmbeddingFunc",
    "EmbeddingRouter",
    "HandlerDescriptor",
]

# ── Types ───────────────────────────────────────────────────────────

Embedding = list[float]
"""A dense vector representation of text."""


# ── Constants ───────────────────────────────────────────────────────

MINIMUM_THRESHOLD: float = 0.0
MAXIMUM_THRESHOLD: float = 1.0
DEFAULT_THRESHOLD: float = 0.3
DEFAULT_TOP_K: int = 5
NO_MATCH_CONFIDENCE: float = 0.0
NO_MATCH_INTENT: str = "unknown"
MATCH_INTENT: str = "semantic"


# ── Protocols ───────────────────────────────────────────────────────


class EmbeddingFunc(Protocol):
    """Async callable that converts text into an embedding vector.

    Implementations may call an external API (OpenAI, Cohere) or run
    a local model.  The returned vector length must be consistent
    across calls.
    """

    async def __call__(self, text: str) -> Embedding:
        """Embed a single text string.

        Args:
            text: The input text to embed.

        Returns:
            A dense float vector representing the text.
        """
        ...


# ── Value objects ───────────────────────────────────────────────────


@dataclass(frozen=True)
class HandlerDescriptor:
    """A handler registered with the embedding router.

    Attributes:
        name: Handler name matching a registry entry.
        description: Human-readable description that was embedded.
        embedding: Pre-computed embedding vector (set at register time).
    """

    name: str
    description: str
    embedding: Embedding


# ── Router ──────────────────────────────────────────────────────────


class EmbeddingRouter:
    """Router using cosine similarity between message and handler descriptions.

    Handlers are registered with descriptions.  At classify time the message
    is embedded and compared against all handler embeddings.  Top-k handlers
    above the confidence threshold are returned, best first.

    Args:
        embed: Async function that converts text to an embedding vector.
        threshold: Minimum cosine similarity to consider a match (0.0-1.0).
        top_k: Maximum number of candidates to return.

    Raises:
        ValueError: If *threshold* is outside [0.0, 1.0] or *top_k* < 1.
    """

    def __init__(
        self,
        embed: EmbeddingFunc,
        *,
        threshold: float = DEFAULT_THRESHOLD,
        top_k: int = DEFAULT_TOP_K,
    ) -> None:
        if not (MINIMUM_THRESHOLD <= threshold <= MAXIMUM_THRESHOLD):
            raise ValueError(
                f"threshold must be between {MINIMUM_THRESHOLD} and "
                f"{MAXIMUM_THRESHOLD}, got {threshold}"
            )
        if top_k < 1:
            raise ValueError(f"top_k must be >= 1, got {top_k}")

        self._embed = embed
        self._threshold = threshold
        self._top_k = top_k
        self._handlers: list[HandlerDescriptor] = []

    async def register(self, name: str, description: str) -> None:
        """Register a handler by embedding its description.

        The description is embedded immediately via the injected embed
        function and stored for later comparison.

        Args:
            name: Handler name (must match a registry entry).
            description: Human-readable description of what the handler does.

        Raises:
            ValueError: If *name* is empty or *description* is blank.
        """
        if not name:
            raise ValueError("Handler name must not be empty")
        if not description.strip():
            raise ValueError("Handler description must not be blank")

        embedding = await self._embed(description)
        descriptor = HandlerDescriptor(
            name=name,
            description=description,
            embedding=embedding,
        )
        self._handlers.append(descriptor)

    async def classify(self, message: str, ctx: ExecContext) -> IntentResult:
        """Classify a message by cosine similarity against handler descriptions.

        Embeds the message, computes cosine similarity against every registered
        handler, and returns the top-k handlers above the threshold.

        Args:
            message: Raw user message text.
            ctx: Execution context (carried for protocol conformance and
                observability; not used for routing decisions).

        Returns:
            :class:`~nerva.router.IntentResult` with ranked candidates.
            Returns an empty result if no handlers are registered or
            none exceed the threshold.
        """
        if not message.strip() or not self._handlers:
            return _empty_result()

        query_embedding = await self._embed(message)
        candidates = _rank_handlers(
            query_embedding, self._handlers, self._threshold, self._top_k
        )

        if not candidates:
            return _empty_result()

        return _build_result(candidates)


# ── Pure helpers ────────────────────────────────────────────────────


def _cosine_similarity(a: Embedding, b: Embedding) -> float:
    """Compute cosine similarity between two vectors.

    Delegates to the native Rust implementation when available for
    better performance on large vectors.  Falls back to pure Python
    otherwise.

    Returns 0.0 for zero-length vectors or mismatched dimensions
    rather than raising, so callers never hit division-by-zero.

    Args:
        a: First embedding vector.
        b: Second embedding vector.

    Returns:
        Cosine similarity in the range [-1.0, 1.0], or 0.0 for
        degenerate inputs.
    """
    if len(a) != len(b) or len(a) == 0:
        return 0.0

    if _USE_NATIVE:
        return _native_cosine(a, b)

    return _cosine_similarity_pure(a, b)


def _cosine_similarity_pure(a: Embedding, b: Embedding) -> float:
    """Pure-Python cosine similarity (fallback when native is unavailable).

    Args:
        a: First embedding vector.
        b: Second embedding vector.

    Returns:
        Cosine similarity in the range [-1.0, 1.0], or 0.0 for
        zero-magnitude vectors.
    """
    dot_product = sum(ai * bi for ai, bi in zip(a, b))
    norm_a = math.sqrt(sum(ai * ai for ai in a))
    norm_b = math.sqrt(sum(bi * bi for bi in b))

    if norm_a == 0.0 or norm_b == 0.0:
        return 0.0

    return dot_product / (norm_a * norm_b)


def _rank_handlers(
    query_embedding: Embedding,
    handlers: list[HandlerDescriptor],
    threshold: float,
    top_k: int,
) -> list[HandlerCandidate]:
    """Score all handlers and return the top-k above the threshold.

    Args:
        query_embedding: The embedded user message.
        handlers: All registered handler descriptors.
        threshold: Minimum similarity to include.
        top_k: Maximum number of candidates to return.

    Returns:
        Up to *top_k* :class:`~nerva.router.HandlerCandidate` objects,
        sorted by score descending.
    """
    scored: list[tuple[float, HandlerDescriptor]] = []

    for handler in handlers:
        similarity = _cosine_similarity(query_embedding, handler.embedding)
        # Clamp to [0.0, 1.0] — negative cosine similarity means
        # the vectors point in opposite directions, treat as no match.
        clamped = max(0.0, min(1.0, similarity))
        if clamped >= threshold:
            scored.append((clamped, handler))

    scored.sort(key=lambda pair: pair[0], reverse=True)

    return [
        HandlerCandidate(
            name=descriptor.name,
            score=score,
            reason=f"Cosine similarity {score:.4f} with '{descriptor.description}'",
        )
        for score, descriptor in scored[:top_k]
    ]


def _build_result(candidates: list[HandlerCandidate]) -> IntentResult:
    """Build an IntentResult from ranked candidates.

    Confidence is set to the score of the best candidate.

    Args:
        candidates: Non-empty list of ranked handler candidates.

    Returns:
        IntentResult with semantic intent and the given candidates.
    """
    raw_scores = {c.name: c.score for c in candidates}
    return IntentResult(
        intent=MATCH_INTENT,
        confidence=candidates[0].score,
        handlers=candidates,
        raw_scores=raw_scores,
    )


def _empty_result() -> IntentResult:
    """Build an empty IntentResult when nothing matched.

    Returns:
        IntentResult with zero confidence and no handlers.
    """
    return IntentResult(
        intent=NO_MATCH_INTENT,
        confidence=NO_MATCH_CONFIDENCE,
        handlers=[],
    )
